content/en/blog/_posts/2023-07-18-swap-beta1-graduation/index.md

```diff
@@ -0,0 +1,231 @@
+memory request, the node's total physical memory and the total amount of swap memory on
+the node that is available for use by Pods.
+
+It is important to note that, for containers within Burstable QoS Pods, it is possible to
+opt-out of swap usage by specifying memory requests that are equal to memory limits.
+Containers configured in this manner will not have access to swap memory.
+
+## How does it work?
+There are a number of possible ways that one could envision swap use on a node.
+When swap is already provisioned and available on a node,
+[we have proposed](https://github.com/kubernetes/enhancements/blob/9d127347773ad19894ca488ee04f1cd3af5774fc/keps/sig-node/2400-node-swap/README.md#proposal)
+the kubelet should be able to be configured such that:
+- It can start with swap on.
+- It will direct the Container Runtime Interface to allocate zero swap memory
+  to Kubernetes workloads by default.
+- You can configure the kubelet to specify swap utilization for the entire
+  node.
+
+Swap configuration on a node is exposed to a cluster admin via the
+[`memorySwap` in the KubeletConfiguration](/docs/reference/config-api/kubelet-config.v1).
+As a cluster administrator, you can specify the node's behaviour in the
+presence of swap memory by setting `memorySwap.swapBehavior`.
+
+The kubelet employs the CRI (container runtime interface) API to direct the CRI to
+configure specific cgroup v2 parameters (such as `memory.swap.max`) in a manner that will
+enable the desired swap configuration for a container. The CRI is then responsible to
+write these settings to the container-level cgroup.
+
+## How can I monitor swap?
+A notable deficiency in the Alpha version was the inability to monitor and introspect swap
+usage. This issue has been addressed in the Beta version introduced in Kubernetes 1.28, which now
+provides the capability to monitor swap usage through several different methods.
+
+The beta version of kubelet now collects
+[node-level metric statistics](/docs/reference/instrumentation/node-metrics/),
+which can be accessed at the /metrics/resource and /stats/summary endpoints. This allows users to
+monitor swap usage and remaining swap memory when using LimitedSwap. Additionally, a
+machine_swap_bytes metric has been added to cadvisor to show the total physical swap capacity of the
+machine.
+
+## Caveats
+Having swap available on a system reduces predictability. Swap's performance is
+worse than regular memory, sometimes by many orders of magnitude, which can
+cause unexpected performance regressions. Furthermore, swap changes a system's
+behaviour under memory pressure, and applications cannot directly control what
+portions of their memory usage are swapped out. Since enabling swap permits
+greater memory usage for workloads in Kubernetes that cannot be predictably
+accounted for, it also increases the risk of noisy neighbours and unexpected
+packing configurations, as the scheduler cannot account for swap memory usage.
```

Comment:
OK, sure, but please have a read of my article https://www.scalefactory.com/blog/2019/06/21/to-swap-or-not/ - or another like it - and have a think about whether paging still happens when swap isn't enabled.

_Disabling_ swap may still put applications in a situation where they cannot directly control what
portions of their mapped memory usage need refreshing into RAM.

Reply:
Hey @sftim!
I've read your article and it was very interesting! Thanks for sharing that!

You're right, paging can still occur when swap is disabled. However, as I see it, this topic is out of scope for this blog-post that deals specifically with swap. The behavior you're describing is, as you said, unrelated to swap as it can occur when swap is not provisioned at all.

Do you think that some information should be added here? If so, what do you think would be valuable to add in this context?

Reply:
I recommend you reword
>  Swap's performance is worse than regular memory, sometimes by many orders of magnitude, which can cause unexpected performance regressions. Furthermore, swap changes a system's behaviour under memory pressure, and applications cannot directly control what portions of their memory usage are swapped out

to avoid implying that with swap off, applications get to control what parts of their virtual address space are in RAM.